Overlapping stats request when switch reconnects

Hi,

When a switch disconnects in the middle of a FlowStats request, we will see Overalapping stats request messages being logged for a few cycles before it discards the old request and send a new one.

This can lead to a delayed Flow Consistency routine, as well as Kytos Stats delayed reports, among others. Another impact we are actually seeing is the End to End tests eventually failing in the following test case:

```
rerun: 0
tests/test_e2e_60_of_multi_table.py::TestE2EOfMultiTable::test_010_delete_miss_flow: 2023-09-18,07:08:29.768108 - 2023-09-18,07:08:54.694563
self = <tests.test_e2e_60_of_multi_table.TestE2EOfMultiTable object at 0x7f2759f5ad30>

   def test_010_delete_miss_flow(self):
       """Delete a miss flow so is recreated"""
       pipeline = {
           "multi_table": [
               {
                   "table_id": 0,
                   "description": "First table for miss flow entry",
                   "table_miss_flow": {
                       "priority": 0,
                       "instructions": [{
                           "instruction_type": "goto_table",
                           "table_id": 1
                       }],
                       "match": {"dl_vlan": 101}
                   },
               },
               {
                   "table_id": 1,
                   "description": "Second table for coloring",
                   "napps_table_groups": {
                       "coloring": ["base"],
                       "of_lldp": ["base"],
                       "mef_eline": ["evpl", "epl"],
                   },
               },
           ]
       }

       # Add pipeline
       api_url = f"{KYTOS_API}{OF_MULTI_TABLE_API}"
       response = requests.post(api_url, json=pipeline)
       data = response.json()
       assert response.status_code == 201, response.text
       assert 'id' in data

       # Enabled pipeline
       api_url = f"{KYTOS_API}{OF_MULTI_TABLE_API}/{data['id']}/enable"
       response = requests.post(api_url)
       assert response.status_code == 200, response.text
       time.sleep(10)

       s1 = self.net.net.get('s1')
       flows_s1 = s1.dpctl('dump-flows')
       assert 'actions=resubmit(,1)' in flows_s1

       # Delete all flows from switch 1
       s1.dpctl('del-flows')
       self.net.reconnect_switches()

       time.sleep(10)

       flows_s1 = s1.dpctl('dump-flows')
     assert 'actions=resubmit(,1)' in flows_s1
E       AssertionError: assert 'actions=resubmit(,1)' in ' cookie=0xab00000000000001, duration=13.738s, table=1, n_packets=0, n_bytes=0, priority=50000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'

tests/test_e2e_60_of_multi_table.py:277: AssertionError

``` 


Looking into the Kytos logs we can see why this is happening: 

```
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,093 (MainThread) kytos.core.atcp_server:INFO atcp_server:182:  Connection lost with client 127.0.0.1:51402. Reason: Request closed by client
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,151 (MainThread) kytos.core.atcp_server:INFO atcp_server:182:  Connection lost with client 127.0.0.1:51406. Reason: Request closed by client
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,203 (MainThread) kytos.core.atcp_server:INFO atcp_server:182:  Connection lost with client 127.0.0.1:51404. Reason: Request closed by client
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,253 (MainThread) kytos.core.atcp_server:INFO atcp_server:131:  New connection from 127.0.0.1:51436
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,318 (MainThread) kytos.core.atcp_server:INFO atcp_server:131:  New connection from 127.0.0.1:51438
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,366 (thread_pool_sb_0) kytos.napps.kytos/of_core:INFO main:158:  Connection ('127.0.0.1', 51436), Switch 00:00:00:00:00:00:00:01: OPENFLOW HANDSHAKE COMPLETE
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,537 (thread_pool_sb_4) kytos.napps.kytos/of_core:INFO main:158:  Connection ('127.0.0.1', 51438), Switch 00:00:00:00:00:00:00:02: OPENFLOW HANDSHAKE COMPLETE
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,649 (MainThread) kytos.core.atcp_server:INFO atcp_server:131:  New connection from 127.0.0.1:51440
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,773 (thread_pool_sb_2) kytos.napps.kytos/of_core:INFO main:82:  Overlapping stats request: switch 00:00:00:00:00:00:00:01 flows_xid 3246955587 ports_xid 2515420549
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,869 (AnyIO worker thread) kytos.napps.kytos/flow_manager:INFO utils:196:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: add, force: False,  flows[0, 1]: [{'actions': [{'action_type': 'output', 'port': 4294967293}], 'table_id': 1, 'owner': 'of_lldp', 'table_group': 'base', 'priority': 50000, 'cookie': 12321848580485677057, 'match': {'dl_type': 35020, 'dl_vlan': 3799}}]
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,931 (MainThread) uvicorn.access:INFO httptools_impl:506:  127.0.0.1:46862 - "POST /api/kytos/flow_manager/v2/flows/00%3A00%3A00%3A00%3A00%3A00%3A00%3A01 HTTP/1.1" 202
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,922 (AnyIO worker thread) kytos.napps.kytos/flow_manager:INFO utils:196:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:02, command: add, force: False,  flows[0, 1]: [{'actions': [{'action_type': 'output', 'port': 4294967293}], 'table_id': 1, 'owner': 'of_lldp', 'table_group': 'base', 'priority': 50000, 'cookie': 12321848580485677058, 'match': {'dl_type': 35020, 'dl_vlan': 3799}}]
Sep 18 07:08:40 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:40,981 (MainThread) uvicorn.access:INFO httptools_impl:506:  127.0.0.1:46864 - "POST /api/kytos/flow_manager/v2/flows/00%3A00%3A00%3A00%3A00%3A00%3A00%3A02 HTTP/1.1" 202
Sep 18 07:08:41 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:41,049 (thread_pool_sb_2) kytos.napps.kytos/of_core:INFO main:158:  Connection ('127.0.0.1', 51440), Switch 00:00:00:00:00:00:00:03: OPENFLOW HANDSHAKE COMPLETE
Sep 18 07:08:41 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:41,177 (AnyIO worker thread) kytos.napps.kytos/flow_manager:INFO utils:196:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:03, command: add, force: False,  flows[0, 1]: [{'actions': [{'action_type': 'output', 'port': 4294967293}], 'table_id': 1, 'owner': 'of_lldp', 'table_group': 'base', 'priority': 50000, 'cookie': 12321848580485677059, 'match': {'dl_type': 35020, 'dl_vlan': 3799}}]
Sep 18 07:08:41 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:41,195 (MainThread) uvicorn.access:INFO httptools_impl:506:  127.0.0.1:46866 - "POST /api/kytos/flow_manager/v2/flows/00%3A00%3A00%3A00%3A00%3A00%3A00%3A03 HTTP/1.1" 202
Sep 18 07:08:41 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:41,794 (thread_pool_sb_6) kytos.napps.kytos/of_core:INFO main:82:  Overlapping stats request: switch 00:00:00:00:00:00:00:03 flows_xid 2871009987 ports_xid 18022052
Sep 18 07:08:47 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:47,294 (Thread-95) kytos.napps.kytos/of_core:INFO main:82:  Overlapping stats request: switch 00:00:00:00:00:00:00:01 flows_xid 3246955587 ports_xid 2515420549
Sep 18 07:08:47 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:47,651 (Thread-96) kytos.napps.kytos/of_core:INFO main:82:  Overlapping stats request: switch 00:00:00:00:00:00:00:03 flows_xid 2871009987 ports_xid 18022052
Sep 18 07:08:54 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:54,314 (Thread-98) kytos.napps.kytos/of_core:INFO main:82:  Overlapping stats request: switch 00:00:00:00:00:00:00:01 flows_xid 3246955587 ports_xid 2515420549
Sep 18 07:08:54 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:54,671 (Thread-99) kytos.napps.kytos/of_core:INFO main:82:  Overlapping stats request: switch 00:00:00:00:00:00:00:03 flows_xid 2871009987 ports_xid 18022052
Sep 18 07:08:54 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:54,859 (MainThread) uvicorn.error:INFO on:67:  Waiting for application shutdown.
Sep 18 07:08:54 runner-rmtk68xb-project-107-concurrent-0 2023-09-18 07:08:54,860 (MainThread) uvicorn.error:INFO on:78:  Application shutdown complete.
```

In summary: after the switch disconnects (while a flow stats requests was pending from reply), the of_core keeps waiting for the reply of the old request, until a point where it eventually clear the existing request and send a new one. Switches seems to discard flow stats requests that were sent before the switch reconnects to controller, so I believe it is safe to clear flow stats request after a switch reconnects.

Ccing @viniarck to also hear his opinion on that.

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -6,6 +6,11 @@ All notable changes to the of_core NApp will be documented in this file.
 [UNRELEASED] - Under development
 ********************************
 
+Fixed
+=====
+- Multipart replies clean up now happens before connection gets established to be safer
+
+
 [2023.1.0] - 2023-06-05
 ***********************
 

--- a/main.py
+++ b/main.py
@@ -147,6 +147,7 @@ class Main(KytosNApp):
         version_utils = self.of_core_version_utils[connection.protocol.version]
         switch = version_utils.handle_features_reply(self.controller, event)
         switch.update_lastseen()
+        self.pop_multipart_replies(switch)
 
         if (connection.is_during_setup() and
                 connection.protocol.state == 'waiting_features_reply'):
@@ -544,14 +545,6 @@ class Main(KytosNApp):
         """Close the connection upon hello failure."""
         event.destination.close()
         log.debug("Connection %s: Connection closed.", event.destination.id)
-
-    @alisten_to(".*.connection.lost")
-    async def on_connection_lost(self, event) -> None:
-        """On connection_lost event."""
-        switch = event.content["source"].switch
-        if not switch:
-            return
-        self.pop_multipart_replies(switch)
 
     def pop_multipart_replies(self, switch) -> None:
         """Pop multipart replies."""

--- a/tests/unit/test_main.py
+++ b/tests/unit/test_main.py
@@ -370,26 +370,6 @@ class TestNApp:
         await napp.aemit_message_out(mock_connection, mock_message)
         mock_aemit_message_out.assert_called()
 
-    async def test_on_connection_lost(self, napp) -> None:
-        """Test on_connection_lost."""
-        dpid = "1"
-        event, switch = MagicMock(), MagicMock(id=dpid)
-        event.content["source"].switch = switch
-        napp._multipart_replies_xids[dpid] = {"flows": 2, "ports": 3}
-        napp._multipart_replies_flows[dpid] = [MagicMock()]
-        napp._multipart_replies_ports[dpid] = [MagicMock()]
-        await napp.on_connection_lost(event)
-        assert dpid not in napp._multipart_replies_xids
-        assert dpid not in napp._multipart_replies_flows
-        assert dpid not in napp._multipart_replies_ports
-
-        # To also cover the early return
-        event = MagicMock()
-        event.content["source"].switch = None
-        napp.pop_multipart_replies = MagicMock()
-        await napp.on_connection_lost(event)
-        napp.pop_multipart_replies.assert_not_called()
-
     async def test_on_openflow_connection_error(self, napp) -> None:
         """Test on_openflow_connection_error."""
         dpid = "1"
@@ -559,6 +539,8 @@ class TestMain(TestCase):
          mock_send_set_config_v0x04) = args
         mock_buffers_put = MagicMock()
         self.napp.controller._buffers.app.put = mock_buffers_put
+        dpid = self.switch_v0x04.connection.switch.dpid
+        self.switch_v0x04.connection.switch.id = dpid
         mock_freply_v0x04.return_value = self.switch_v0x04.connection.switch
         name = 'kytos/of_core.v0x04.messages.in.ofpt_features_reply'
         self.switch_v0x04.connection.state = ConnectionState.SETUP
@@ -567,6 +549,13 @@ class TestMain(TestCase):
         event = get_kytos_event_mock(name=name, content=content)
         count = self.switch_v0x04.connection.switch.update_lastseen.call_count
         assert count == 0
+
+        # To simulate a few existing multipart replies for extra test cov
+        self.napp._multipart_replies_xids[dpid] = {"flows": 2, "ports": 3}
+        self.napp._multipart_replies_flows[dpid] = [MagicMock()]
+        self.napp._multipart_replies_ports[dpid] = [MagicMock()]
+        self.napp._multipart_replies_tables[dpid] = [MagicMock()]
+
         self.napp.handle_features_reply(event)
         count = self.switch_v0x04.connection.switch.update_lastseen.call_count
         assert count == 1
@@ -575,6 +564,12 @@ class TestMain(TestCase):
             self.napp.controller, self.switch_v0x04.connection.switch)
         mock_send_set_config_v0x04.assert_called_with(
             self.napp.controller, self.switch_v0x04.connection.switch)
+
+        # Make sure it was cleaned up when handling features reply
+        assert dpid not in self.napp._multipart_replies_xids
+        assert dpid not in self.napp._multipart_replies_flows
+        assert dpid not in self.napp._multipart_replies_ports
+        assert dpid not in self.napp._multipart_replies_tables
 
         mock_buffers_put.assert_called()
 

